nexus/src/app/sagas/region_replacement_start.rs

```diff
@@ -450,76 +449,57 @@ async fn srrs_get_old_region_address(
     let osagactx = sagactx.user_data();
     let params = sagactx.saga_params::<Params>()?;
 
-    // It was a mistake not to record the port of a region in the Region record.
-    // However, we haven't needed it until now! If the Crucible agent is gone
-    // (which it will be if the disk is expunged), assume that the region in the
-    // read/write portion of the volume with the same dataset address (there
-    // should only be one due to the allocation strategy!) is the old region.
+    // Either retrieve the address from the database (because the port was
+    // previously recorded), or attempt grabbing the port from the corresponding
+    // Crucible agent: the sled or disk may not be physically gone, or we may be
+    // running in a test where the allocation strategy does not mandate distinct
+    // sleds.
 
-    let opctx = crate::context::op_context_for_saga_action(
-        &sagactx,
-        &params.serialized_authn,
-    );
-
-    let db_region = osagactx
-        .datastore()
-        .get_region(params.request.old_region_id)
-        .await
-        .map_err(ActionError::action_failed)?;
-
-    let targets = osagactx
-        .datastore()
-        .get_dataset_rw_regions_in_volume(
-            &opctx,
-            db_region.dataset_id(),
-            db_region.volume_id(),
-        )
+    let maybe_addr = osagactx
+        .nexus()
+        .region_addr(log, params.request.old_region_id)
```

Comment:
This is calling the function defined in `nexus/src/app/crucible.rs`, which tries to access the port and set if if we don't have the value, right?

If this is already doing the work of contacting the agent and back-filling the port, why are we bothering to handle the `match maybe_addr => None` case below? Isn't that redundant with the work in `nexus/src/app/crucible.rs`'s `region_addr` function?

Reply:
In order for region replacement to continue even if the agent is gone, this saga node makes the assumption that if there's a unique target with the dataset's address in the region set, then that's the region's port. It's not redundant in that there's an assumption made, not a lookup.

However, if it's good enough to make this assumption for region replacement, the question then becomes: if there is only one target using the dataset's address, should that target's port be recorded in the database as this region's port? I'm thinking yes.

Reply:
Ack, the structure makes more sense to me. It's basically:

- Backfill the port info if we can
- If we can't, at least confirm that the address is known to be unique
- If we can't do that, bail